stable/jenkins/README.md

```diff
@@ -169,6 +172,28 @@ AdditionalConfig:
       clientURL: testUrl
 ```
 
+## Using Configuration as Code
+
+`Master.ConfigurationAsCode.Enabled` can be used to enable [Jenkins Configuration as Code Plugin](https://github.com/jenkinsci/configuration-as-code-plugin).
+It automatically adds it to `Master.InstallPlugins` unless it's already specified there.
+
+The plugin version being used can be customized via `Master.ConfigurationAsCode.PluginVersion`.
+
+Below is an example how you can customize the system message by providing `Master.ConfigurationAsCode.AdditionalConfig`:
+
+```yaml
+Master:
+  ConfigurationAsCode:
+    Enabled: true
+    AdditionalConfig:
+      testConfig.yaml: |-
+        jenkins:
+          systemMessage: "Jenkins configured automatically by Jenkins Configuration as Code Plugin\n\n"
+```
+
+In contrast to adding xml configuration files these settings are reapplied during every Jenkins restart and when you apply
+the configuration in the plugin ui.
```

Comment:
You're correct. What I wanted to say was that in contrast to changes using the xml file it's safer to use CASC configuration as you have more control over which settings are modified. For example the `config.xml` stores configuration options from multiple plugins. So if you just want to modify the setting of one plugin using the xml approach you are rather screwed if you try to copy and overwrite the file. CASC in my opinion is clearer in that regards. But even there if you decide to manage credentials via configuration as code, then you need to manage all credentials via config as code as other credentials are removed.

So as the judgement is rather opinionated I left if out from the documentation and it now reads:

> Every configuration changes triggers a Jenkins restart during which the settings are reapplied.

Jenkins restart is intended to ensure that the configuration is applied. It achieved by adding a `checksum/casc` annotation in the deployment:

```
      annotations:
        checksum/config: {{ include (print $.Template.BasePath "/config.yaml") . | sha256sum }}
        {{- if .Values.Master.ConfigurationAsCode.Enabled }}
        checksum/casc: {{ include (print $.Template.BasePath "/configmap-jenkins-config-as-code.yaml") . | sha256sum }}
        {{- end}}
```

It's actually possible to watch for config map changes e.g. via sidecar container, but I don't know how to trigger a config reload in Jenkins. The plugin offers a rest endpoint `/configuration-as-code/reload` to trigger a reload using a POST request. The issue is that the endpoint needs authentication. So one could trigger it via:

`curl -u "user:pass" -X POST http://localhost:8080/configuration-as-code/reload`

I thought about mounting the jenkins admin credentials into a sidecar container and to use those, but decided against it because:
- first of all it does not work as one needs a token and not a password (otherwise Jenkins complains about a missing crumb)
- second and even more important it would not work if people configured Jenkins using different authentication methods e.g. Active Directory / LDAP.

Reply:
"Every configuration changes triggers a Jenkins restart" - Hence not sure what this offers beyond [the PR I already raised](https://github.com/helm/charts/pull/9057).  Without config auto-reload, it's moot whether adding an extra configuration method is worth the steeper learning curve.  I've got sidecar functionality working, just need to ensure it can be secure enough.  Should be able to push it in the coming week.

Reply:
@holmesb To be honest I haven't seen your PR before submitting mine. Next time I'll check for existing PRs.

To respond to your question what this offers beyond the exiting PR: The main differences I see are:
- it separates casc configuration from existing configuration by using a new config map
  That's potentially useful if one in the future would like to support reloading config changes without restarts as changes to the `config.yaml` always trigger a restart. For the new config map that behavior could be changed. If https://github.com/jenkinsci/configuration-as-code-plugin/issues/76 is resolved than restarting would no longer be necessary.
- it does not add `configuration-as-code` plugin for people which are not going to use it
- it allows specifying the `configuration-as-code` plugin version without altering other default plugins

Actually I like your description here: https://github.com/helm/charts/pull/9057/files#diff-bd43a39639bf663f4e99c0d186fad542R128

For sure it does not make sense to have two PRs for the same topic. I would like to collaborate with you on the PR. Please let me know what you think about the differences and if they make sense. I would be happy to close this PR and add those changes to your or the other way round. Whatever you prefer.

>  I've got sidecar functionality working, just need to ensure it can be secure enough.

Is it available in your repo? Could you send me a link?